### DISCUSSION TITLE
llvm-strip throws invalid relocation offset when using --strip-debug

### DISCUSSION THREAD

**Discussion by gmarella** [emscripten-core/emscripten#22289] [Q&A] [answered]
Hi,

I am trying to use `llvm-strip` that gets shipped with emscripten to strip the debug information from the static library compiled using `emcc`. It always throws the error `invalid relocation offset` when the strip command is executed. This looked basic when we started but not able to get it working. Any inputs?

```
make debug-strip
cp out/libsimple_thread.a out/libsimple_thread.a.orig.debug
~/repos/emsdk/upstream/bin/llvm-strip --strip-debug out/libsimple_thread.a
~/repos/emsdk/upstream/bin/llvm-strip: error: 'out/libsimple_thread.a': 'simple_thread.o': invalid relocation offset
make: *** [debug-strip] Error 1
```
I also tried removing the debug section from the `simple_thread.o` first and then creating the archive but hit the same error during archive creation step.
```
~/repos/emsdk/upstream/bin/llvm-objcopy --strip-debug ./out/simple_thread.o ./out/simple_thread_lean.o
~/repos/emsdk/upstream/bin/llvm-ar crus libmylib.a out/simple_thread_lean.o
/Users/gmarella/repos/emsdk/upstream/bin/llvm-ar: error: libmylib.a: 'simple_thread_lean.o': invalid relocation offset
```

Code structure is as follows (Gist is also at https://gist.github.com/gmarella/9caa8a88fdd221c1fed4062e5a81c782 )
- Makefile
- src
   - main.cpp
   - simple_thread.cpp
   - simple_thread.h

#### Libary code
simple_thread.cpp
```cpp
#include <iostream>
#include <string>
#include <pthread.h>
#include <unistd.h>

using namespace std;

void* thread_function(void *)
{
    cout << "Inside thread function" << endl;
    while (true)
    {
        cout << "Looping in thread" << endl;
        sleep(1);
    }

    return nullptr;
}
```
simple_thread.h
```cpp
#pragma once

void* thread_function(void *);
```
main.cpp
```cpp

#include <stddef.h>
#include <string.h>
#include <stdint.h>
#include <iostream>

#include <pthread.h>
#include "simple_thread.h"

using namespace std;

void InitThreads()
{
    pthread_t thread;
    pthread_create(&thread, NULL, &thread_function, NULL);
    pthread_join(thread, NULL);
}

int main()
{
    InitThreads();

    return 0;
}
```

Makefile
```
EMSCRIPTEN_BASE=~/repos/emsdk/upstream/emscripten
LLVM_TOOLS_BASE=~/repos/emsdk/upstream/bin

EMCC = $(EMSCRIPTEN_BASE)/emcc
EMAR = $(EMSCRIPTEN_BASE)/emar

OUT_DIR = out
SERVE_DIR = serve
SRC_DIR = src

# Define the source file(s)
MAIN_SRC = $(SRC_DIR)/main.cpp 

LIB_SRC_FILES = simple_thread.cpp
LIB_SRC = $(addprefix $(SRC_DIR)/, $(LIB_SRC_FILES))
LIB_OBJS = $(LIB_SRC_FILES:.cpp=.o)
LIB_NAME = libsimple_thread.a

# Define the HTML file
HTML_OUT = index

# Compiler flags
CFLAGS = -g -pthread -fPIC

LFLAGS = -s WASM=1 --emrun

TARGET = output

all: $(TARGET)

$(OUT_DIR):
	mkdir -p $(OUT_DIR)

$(SERVE_DIR):
	mkdir -p $(SERVE_DIR)

$(LIB_OBJS): $(LIB_SRC) | $(OUT_DIR)
	$(EMCC) $(CFLAGS) -c $(LIB_SRC) -o $(OUT_DIR)/$@

$(LIB_NAME): $(LIB_OBJS)
	$(EMAR) rcs $(OUT_DIR)/$(LIB_NAME) $(OUT_DIR)/$(LIB_OBJS)

$(TARGET): $(MAIN_SRC) $(LIB_NAME) | $(SERVE_DIR)
	$(EMCC) $(CFLAGS) $(LFLAGS) $(MAIN_SRC) $(OUT_DIR)/$(LIB_NAME) -o $(SERVE_DIR)/$(HTML_OUT).html

# Rule to clean the output (for cleanup purposes)
clean:
	rm -rf $(OUT_DIR) $(SERVE_DIR)

debug-strip:
	cp $(OUT_DIR)/$(LIB_NAME) $(OUT_DIR)/$(LIB_NAME).orig.debug
	$(LLVM_TOOLS_BASE)/llvm-strip --strip-debug $(OUT_DIR)/$(LIB_NAME)

```

**Comment by kripken** [comment]
The test we have for `emstrip` (which just runs `llvm-strip`) runs on a wasm file, not a library or an object file:

https://github.com/emscripten-core/emscripten/blob/8c81cac1bbae378bc7dde0c21c99602cbaf452d0/test/test_other.py#L870-L876

@dschuff @sbc100 Perhaps that tool is only meant to be run like that?

In any case @gmarella , as a workaround you can run strip at the very end, on the wasm, and not in the middle. The result should be the same.

**Reply by gmarella** [reply]
Hi @kripken,

Thanks for the reply. Our use case is little different, we build WASM by linking a set of static libraries. For the ease of local development, we want to have a choice of which libraries to consider including the debug symbols for. For unwanted libraries, we want to strip the debug symbols first and then build WASM along with a dedicated dwarf symbol file with debug info only from the required libraries.

**Comment by dschuff** [comment, marked as answer]
This does look like a bug, I created #22291 
The fact that the crash is the same when trying to strip the object file and then add it to the archive after that means that you can't work around the bug by working only on object files, and suggests that the file created by the stripping is corrupted in a way that crashes the object file parser.

**Comment by sbc100** [comment]
Lets close this an followup on #22291
